File: hw1/linkedlist.py
class Node():
	def __init__(self):
		self.cargo = None
		self.next = None

	def __init__(self, newCargo):
		self.cargo = newCargo
		self.next = None

	def getNext(self):
		return self.next

	def getCargo(self):
		return self.cargo

	def insertAfter(self, newItem):
		if type(newItem) == str:
			self.next = Node(newItem)
		elif type(newItem) == Node:
			self.next = newItem

	def __eq__(self, other):
		if other == None:
			return False
		return self.cargo == other.cargo

	def __str__(self):
		strBuffer = ""

		if type(self.cargo) == str:
			strBuffer += "\'"

		strBuffer += self.cargo

		if type(self.cargo) == str:
			strBuffer += "\'"

		return strBuffer

class LinkedList():
	def __init__(self):
		self.head = None

	def isEmpty(self):
		return self.head == None

	def append(self, newCargo):
		if self.isEmpty():
			self.head = Node(newCargo)
		else:
			iterNode = self.head

			while iterNode.getNext() != None:
				iterNode = iterNode.getNext()

			iterNode.insertAfter(newCargo)

	def find(self, item):
		iterNode = self.head
		depth = 0

		while iterNode != None:
			if (type(item) == str and iterNode.getCargo() == item) or (type(item) == Node and iterNode == item):
				return depth

			iterNode = iterNode.getNext()
			depth += 1

		return -1

	def populate(self, fileName, delimiter):
		self.head = None

		fileString = open(fileName, "r").read()

		items = fileString.split(delimiter)

		for item in items:
			self.append(item)

	def __str__(self):
		strBuffer = ""
		strBuffer += "LinkedList: "

		if self.isEmpty():
			strBuffer += "None."

		else:
			strBuffer += "["

			iterNode = self.head
			while iterNode.getNext() != None:
				strBuffer += str(iterNode) + ", "
				iterNode = iterNode.getNext()

			strBuffer += str(iterNode) + "]"


		return strBuffer


def main():
	pass

if __name__ == "__main__":
	main()

File: hw12/linkedlist.js
module.exports = {
  LinkedList: LinkedList
}

var fs = require('fs');


/**
 * read takes in a file name (or path), and returns a Promise that will resolve with
 * the utf-8 encoded contents of the file
 * 
 * @param: fileName: The path to the file to be read in
*/
function read(fileName) {
  return new Promise(function promiseRead(resolve, reject) {
    fs.readFile(fileName, { encoding: 'utf8' }, function handleResponse(err, data) {
      if (err) { reject(Error(err)); }
      else { resolve(data); }
    });
  });
}

/**
 * Default constructor for the Node class
*/
function Node() {
  this.value = null;
  this.next = null;
}

/**
 * Parameterized constructor for the Node class
 * Takes in a value to store in this node
 * 
 * @param: newValue: The value to store
*/
function Node(newValue) {
  this.value = newValue;
  this.next = null;
}

/**
 * Default constructor for the LinkedList class
*/
function LinkedList() {
  this.head = null;
  this.tail = null;
}

/**
 * Append (or push) method for the LinkedList class
 * Takes one value to store in a new node, appended as the last element of the list
 * 
 * @param: newValue: The new value to append to the list
*/
LinkedList.prototype.append = function append(newValue) {
  if (this.head == null) {
    this.head = new Node(newValue);
    this.tail = this.head;
  }

  else {
    this.tail.next = new Node(newValue);
    this.tail = this.tail.next;
  }
}

/**
 * Copies the contents of an array into the LinkedList object
 * 
 * @param: arrayData: The array of data to copy over
*/
LinkedList.prototype.insertIntoList = function insertIntoList(arrayData) {
  for (var i in arrayData) {
    this.append(arrayData[i]);
  }
}

/**
 * Populates the LinkedList from the file provided by the fileName parameter.
 * The format of the given file is assumed to be a delimiter-separated list of strings.
 * 
 * @param: fileName: The name of the file to load in
 * @param: delimiter: The string that separates the values of the list, stored in the file
*/
LinkedList.prototype.populate = function populate(fileName, delimiter) {
  return read(fileName).then(function splitOnDelimiter(fileData) {
    return fileData.split(delimiter);
  }).then(this.insertIntoList.bind(this))
  .catch(function handleError(err) {
    console.error(err.stack);
    throw err;
  });
}

/**
 * Searches the LinkedList for a stored value, returning the index at which the value
 * was found. Returns -1 if not found.
 * 
 * @param: value: The value to find in the LinkedList
 * 
 * @return: The index at which the value was found. -1 if not.
*/
LinkedList.prototype.find = function find(value) {
  var index = 0;
  itNode = this.head;

  while (itNode != null) {
    if (itNode.value == value) {
      break;
    }

    ++index;
    itNode = itNode.next;
  }

  if (itNode == null) {
    index = -1;
  }

  return index;
}
LinkedList.prototype.search = LinkedList.prototype.find;

/**
 * Converts the LinkedList to a string format, similar to JSON. Useful for debugging output
 * when used with console.log().
 * 
 * @return: The string representation of the string.
*/
LinkedList.prototype.toString = function toString() {
  var itNode = this.head;

  var strBuffer = '';

  while (itNode != null) {
    if (typeof itNode.value == "string") {
      strBuffer += '"' + itNode.value + '"' + ', ';
    }
    else {
      strBuffer += itNode.value + ', ';
    }

    itNode = itNode.next;
  }

  return '[' + strBuffer.slice(0, -2) + ']';
}